Create a discussion on versioning

jeanas opened the pull request:
Move content on versioning schemes from
distributing-packages-using-setuptools.rst to a new discussion on versioning, and update it.

This is the continuation of #1450.

<!-- readthedocs-preview python-packaging-user-guide start -->
----
📚 Documentation preview 📚: https://python-packaging-user-guide--1455.org.readthedocs.build/en/1455/discussions/versioning/

<!-- readthedocs-preview python-packaging-user-guide end -->

chrysle commented:
Looks good, thanks! Probably mention a few projects that comply with the different versioning systems?

webknjaz commented on source/discussions/versioning.rst:
```diff
@@ -0,0 +1,139 @@
+
+.. code-block:: text
+
+   1.2.0.dev1+hg.5.b11e5e6f0b0b  # 5th VCS commit since 1.2.0.dev1 release
```
Can we show versions that `setuptools-scm` makes too? I feel like this is something that people would see in the wild more frequently.

webknjaz commented on source/discussions/versioning.rst:
```diff
@@ -0,0 +1,139 @@
+particular release is given just the version number.
+
+Calendar version numbers typically take the form *year.month* (for example,
+23.10 for December 2023).
```
Did you mean to use "October" or "12"?
```suggestion
23.10 for October 2023).
```

webknjaz commented on source/discussions/versioning.rst:
```diff
@@ -0,0 +1,139 @@
+  1.2.0b1     # Beta Release
+  1.2.0rc1    # Release Candidate
+  1.2.0       # Final Release
+  1.2.0.post1 # Post Release
```
How about showing off something more complex? Like a pre-release of a dev-release's post-release?

Also, can we include examples using the epoch version component? https://peps.python.org/pep-0440/#version-epochs

webknjaz commented on source/discussions/versioning.rst:
```diff
@@ -0,0 +1,139 @@
+
+.. code-block:: text
+
+  1.2.0.dev1  # Development release
```
Nit: in RST, we usually match the block indentation with the start of the directive name. So it'd need to start under the `c` letter of `code-block`.

webknjaz commented on source/discussions/versioning.rst:
```diff
@@ -0,0 +1,139 @@
+Two versioning schemes are commonly used for Python packages, semantic
+versioning and calendar versioning.
+
+Semantic versioning is recommended for most new projects.
```
By... whom? It's got a lot of problems (like the maintainer and end-user perception mismatches, especially in the context of locking the envs). I think it'd be more accurate to say "semantic versioning is what people tend to be most familiar with and therefore expect to see", right?

It might be a good idea to link https://hynek.me/articles/semver-will-not-save-you/ somewhere too.

jeanas replied:
I hesitated to include this. I only did so because it's in the current guide. I don't like this recommendation either, and it feels a bit silly given that pip itself, arguably the most high-profile Python packaging project, uses calver.

I'll change it to something softer.

jeanas replied:
I changed the wording to be objective ("most projects do this and that").

> It might be a good idea to link https://hynek.me/articles/semver-will-not-save-you/ somewhere too.

I don't feel very comfortable linking to a blog post on an official guide; do you think it's okay?

chrysle replied:
> I don't feel very comfortable linking to a blog post on an official guide; do you think it's okay?

That's also been done in the new _Is setup.py deprecated?_ guide. I personally don't have any reservations.

jeanas replied:
Done; I've also linked a couple other blog posts by packaging tool maintainers.

webknjaz commented on source/discussions/versioning.rst:
```diff
@@ -0,0 +1,139 @@
+Semantic versioning vs. calendar versioning
+===========================================
+
+A versioning scheme is a way to interpret version numbers of a package, and to
```
Maybe, "formalized way to interpret semantics of the version segments"?

webknjaz commented on source/discussions/versioning.rst:
```diff
@@ -0,0 +1,139 @@
+Other schemes
+-------------
+
+Serial versioning refers to the simplest possible versioning scheme, which
```
Can we link some external resource?

jeanas replied:
I'm not sure we're going to find such a resource since it seems virtually nobody advocates this kind of scheme...

webknjaz commented on source/discussions/versioning.rst:
```diff
@@ -0,0 +1,139 @@
+* Zero or more beta releases, denoted with a ".bN" suffix,
+* Zero or more release candidates, denoted with a ".rcN" suffix.
+
+Pip and other modern Python package installers ignore pre-releases by default
```
It's not very accurate — if a pre-release is met in the user-requests or in the dependency declarations, that flags that the resolver should consider pre-releases for those specific projects, regardless of the explicit opt-in.

webknjaz commented on source/guides/distributing-packages-using-setuptools.rst:
```diff
@@ -288,125 +288,6 @@ points (see below).
 
 
 
-
```
Can we link the new document from in place of the removed content?

webknjaz commented on source/discussions/index.rst:
```diff
@@ -8,6 +8,7 @@ specific topic. If you're just trying to get stuff done, see
 .. toctree::
    :maxdepth: 1
 
+   versioning
```
Should this be the first item in the ToC? Thoughts?

jeanas replied:
Not sure. It was a gut feeling, feel free to propose another sorting.

jeanas replied:
@webknjaz Friendly ping?

webknjaz replied:
Let's leave it be, then. Marking as resolved.

webknjaz commented:
@jeanas the first round of review complete.

jeanas commented:
> Looks good, thanks! Probably mention a few projects that comply with the different versioning systems?

Good idea; done. (I've used Sphinx, NumPy and pip.)

jeanas commented:
Thanks for the detailed comments! I've revised much of the page, please take another look.

chrysle commented:
Could you add an own paragraph for [ZeroVer](https://0ver.org/)?

jeanas commented:
> Could you add an own paragraph for [ZeroVer](https://0ver.org/)?

It's a satire, not a serious thing :)

chrysle commented:
I know, it would be to warn users. Maybe that's stupid.

sinoroc commented:
The comments on this PR seem to show that we are venturing into territory where things are much less clear-cut and I wonder whether there is a risk people come to complain because we "bless" things that they don't like. Not sure you know what I mean and maybe it is unwarranted worry from my side, I just wanted to mention it. I do not have any suggestion to offer. In the end, since it is a "discussion", it might be more forgiving...

jeanas commented:
@sinoroc Sorry, I don't understand your comment at all. What “things” are less clear-cut in the new document?

sinoroc commented:
> I don't understand your comment at all.

Fair enough :D

I think my comment can be ignored. I will hide it as well as this one.

webknjaz commented:
> I wonder whether there is a risk people come to complain because we "bless" things that they don't like

That's literally the story of the Python Packaging in the nutshell 😂

webknjaz commented:
I think adding a 0ver remark might be fun, just don't pass judgement..

jeanas commented:
- Added references to blog posts on SemVer's weaknesses
- Added link to https://0ver.org

I think the only unresolved issue right now is https://github.com/pypa/packaging.python.org/pull/1455#discussion_r1429250729

sinoroc commented on source/discussions/versioning.rst:
```diff
@@ -0,0 +1,176 @@
+release candidates, final release. Pip and other modern Python package
+installers ignore pre-releases by default when deciding which versions of
+dependencies to install, unless explicitly requested (e.g., with
+``pip install pkg==1.1a3``).
```
What about `pip install --pre pkg`? Honestly I never use it so I am not sure what it does, but I know it exists.

chrysle replied:
This flag [installs](https://pip.pypa.io/en/stable/cli/pip_install/#pre-release-versions) release candidates and beta versions. Should be combined with `--upgrade`, I think.

webknjaz replied:
@chrysle no, `--upgrade` is orthogonal. `--pre` essentially allows the dependency resolve to consider pre-releases instead of rejecting them right away. It's global and allows transitive pre-releases.

But using `pip install pkg==1.1a3` allows pre-releases implicitly just for `pkg`, because it's an explicit user request, but not other packages (transitive or otherwise) — it's not global.

There's also a corner case that PEP 440 demands but pip doesn't (yet) implement — when the dependency resolution can only be satisfied with pre-releases, the resolver is supposed to allow inevitable pre-releases in order to succeed. Only one sub-case is implemented — if `pkg` *only* has pre-releases published on the PyPI and zero stable releases, `pip` will allow pre-releases for that project too.

`--upgrade` is only significant when the installed version of a package in the tree satisfies the constraints the resolver encountered, but the user wants to force the latest version of all the projects in the tree anyway.

sinoroc commented on source/discussions/versioning.rst:
```diff
@@ -0,0 +1,176 @@
+increment the major number when the incompatibility is high, or to signal a
+shift in the project, rather than for any tiny incompatibility
+[#semver-strictness]_.
+
```
Somewhere in there I would like to read that incrementing the major version number does not imply incompatible change. Sometimes projects increment their major version number for vanity purposes although the library is still 100% backwards compatible. This detail becomes important when dependent projects put unnecessary upper caps on a dependency, assuming that the next major release will be incompatible, which there is have absolutely no way of knowing until the next major release does actually happen.

jeanas replied:
Done.

webknjaz commented on source/discussions/versioning.rst:
```diff
@@ -0,0 +1,178 @@
+- A post-release of an alpha release (possible, but discouraged): 1.2.0a1.post1
+- A simple version with only two components: 23.12
+- A simple version with just one component (possible, but discouraged): 42
+- A version with an epoch: 1!1.0
```
Can we mark all the versions with double backticks? It may be weird to read things like this as normal text.

jeanas replied:
Done.

webknjaz replied:
Marking as resolved.

webknjaz commented on source/discussions/versioning.rst:
```diff
@@ -0,0 +1,178 @@
+- A beta release: 1.2.0b1
+- A release candidate: 1.2.0rc1
+- A post-release: 1.2.0.post1
+- A post-release of an alpha release (possible, but discouraged): 1.2.0a1.post1
```
Is it? Based on what?

jeanas replied:
This is from the spec:

https://packaging.python.org/en/latest/specifications/version-specifiers/#post-releases

webknjaz replied:
Okay, I see in the spec that this is noted. Resolved.

webknjaz commented on source/discussions/versioning.rst:
```diff
@@ -0,0 +1,178 @@
+- A post-release: 1.2.0.post1
+- A post-release of an alpha release (possible, but discouraged): 1.2.0a1.post1
+- A simple version with only two components: 23.12
+- A simple version with just one component (possible, but discouraged): 42
```
Same — what's the source of discouragement?

webknjaz replied:
I don't see that the spec discourages this. It simply mentions that 2-3 segment version numbers are more common.
So I'd drop that statement, unless there's another information source to support this claim.

jeanas replied:
Ok, done.

webknjaz replied:
Marking as resolved.

webknjaz commented on source/discussions/versioning.rst:
```diff
@@ -0,0 +1,178 @@
+.. _semver-brett-cannon: https://snarky.ca/why-i-dont-like-semver/
+.. _semver-donald-stufft: https://caremad.io/posts/2016/02/versioning-software/
+.. _semver-hynek-schlawack: https://hynek.me/articles/semver-will-not-save-you/
+.. _setuptools-scm: https://setuptools-scm.readthedocs.io
```
use intersphinx, maybe?

jeanas replied:
We can't, because these docs are made with MkDocs.

webknjaz replied:
Ah, I forgot. Well, then. Marking as resolved.

webknjaz commented on source/discussions/versioning.rst:
```diff
@@ -0,0 +1,178 @@
+-------------------
+
+The idea of *semantic versioning* (or SemVer) is to use 3-part version numbers,
+*major.minor.maintenance*, where the project author increments:
```
Where is *maintenance* coming from? https://semver.org defines the third component as *patch*.

jeanas replied:
Changed

webknjaz replied:
Marking as resolved.

webknjaz commented on source/discussions/versioning.rst:
```diff
@@ -0,0 +1,178 @@
+Semantic versioning vs. calendar versioning
+===========================================
+
+A versioning scheme is a formalized way to interpret the segments of a version
```
Let's emphasize that the interpretation is a reflection of the maintainer's view which is not necessarily how the end-users would perceive their version bump choice.

And also, it's important to mention that there are exceptions, like security releases that break the versioning scheme assumptions. As in patch releases having security fixes that are inevitably breaking changes by nature.

jeanas replied:
I'm not sure how to phrase this and if you want to add something or to rephrase the paragraph below about "strict" vs "loose" semver. Could you perhaps propose a wording or give more details on what you envision?

webknjaz replied:
Here you go https://github.com/pypa/packaging.python.org/pull/1455/files#r1468739775

webknjaz commented on source/discussions/versioning.rst:
```diff
@@ -0,0 +1,178 @@
+For those projects that do use strict semantic versioning, this approach allows
+users to make use of :ref:`compatible release version specifiers
+<version-specifiers-compatible-release>`, with the ``~=`` operator.  For
+example, ``name ~= X.Y`` is roughly equivalent to ``name >= X.Y, == X.*``, i.e.,
```
Not roughly, exactly equivalent:
```suggestion
example, ``name ~= X.Y`` is equivalent to ``name >= X.Y, == X.*``, i.e.,
```

webknjaz commented:
@jeanas I stumbled upon an article by Seth. Maybe it could be linked from somewhere too: https://sethmlarson.dev/pep-440.

webknjaz commented on source/discussions/versioning.rst:
```diff
@@ -0,0 +1,181 @@
+number, and to decide which should be the next version number for a new release
+of a package. Two versioning schemes are commonly used for Python packages,
+semantic versioning and calendar versioning.
+
```
```suggestion

.. caution::

   The decision which version number to choose is up to a
   project's maintainer. This effectively means that version
   bumps reflect the maintainer's view. That view may differ
   from the end-users' perception of what said formalized
   versioning scheme promises them.

   There are known exceptions for selecting the next version
   number. The maintainers may consciously choose to break the
   assumption that the last version segment only contains
   backwards-compatible changes.
   One such case is when security vulnerability needs to be
   addressed. Security releases often come in patch versions
   but contain breaking changes inevitably.

```

jeanas commented:
@webknjaz I've addressed the latest review.

webknjaz reviewed (approved):
Thanks 🙏
